### PR TITLE
Don't copy parameters on setting a key in libssl

### PR DESCRIPTION
**Pull request by mattcaswell**
Whenever we set a private key in libssl, we first found the certificate that matched the key algorithm. Then we copied the key parameters from the private key into the public key for the certficate before finally checking that the private key matched the public key in the certificate. This makes no sense! Part of checking the private key is to make sure that the parameters match. It seems that this code has been present since SSLeay. Perhaps at some point it made sense to do this - but it doesn't any more.

We remove that piece of code altogether. The previous code also had the undocumented side effect of removing the certificate if the key didn't match. This makes sense if you've just overwritten the parameters in the certificate with bad values - but doesn't seem to otherwise. I've also removed that error logic.

Due to issue #13893, the public key associated with the certificate is always a legacy key. `EVP_PKEY_copy_parameters` will downgrade the "from" key to legacy if the target is legacy, so this means that in libssl all private keys were always downgraded to legacy when they are first set in the `SSL`/`SSL_CTX`. Removing the `EVP_PKEY_copy_parameters code` has the added benefit of removing that downgrade.

Once the above change had been made I found that the tests started to fail. This was due to an unrelated bug that was exposed because we were using provider based keys for the first time in libssl. The `legacy_asn1_ctrl_to_param` implementation of
`ASN1_PKEY_CTRL_DEFAULT_MD_NID` calls `EVP_PKEY_get_default_digest_name() `which returns an `mdname`. Previously we were using `OBJ_sn2nid`/`OBJ_ln2nid` to lookup that name in the OBJ database. However we might get an `mdname` back that only exists in the `namemap`, not in the OBJ database. In that case we need to check the various aliases for the name, to see if one of those matches the name we are looking for.

### CODE REVIEW

File: ssl/ssl_rsa.c [hunk 1]
```diff
@@ -124,26 +124,9 @@ static int ssl_set_pkey(CERT *c, EVP_PKEY *pkey)
...
-
-        if (!X509_check_private_key(c->pkeys[i].x509, pkey)) {
-            X509_free(c->pkeys[i].x509);
-            c->pkeys[i].x509 = NULL;
-            return 0;
-        }
-    }
+    if (c->pkeys[i].x509 != NULL
+            && !X509_check_private_key(c->pkeys[i].x509, pkey))
+        return 0;
```
Comment (levitte) [inline ssl/ssl_rsa.c:129]: I do like this, but vague memory tells me that there were cases when the public key came without parameters, and that in that case, the private key parameters should be used...  not sure if the standards support that, but what goes on in the wild doesn't always correspond to standards...

Comment (mattcaswell) [reply ssl/ssl_rsa.c:129]: These are public keys that come from X509 certificates. Those certificates get sent to clients, that then use those public keys without access to the private key. It is difficult to see how that could possibly work if the public key is not accompanied with its parameters.

**Comment by mattcaswell** [comment]
Ping for review?

File: crypto/evp/p_lib.c [hunk 2]
```diff
@@ -1158,6 +1159,18 @@ int EVP_PKEY_print_params(BIO *out, const EVP_PKEY *pkey,
                       pctx);
 }
 
+static void find_md_nid(const char *mdname, void *data)
```
Comment (t8m) [inline crypto/evp/p_lib.c:1162]: The name of the function is slightly confusing as it does not do the find by itself. Perhaps `mdname2nid` or something similar would be better name?

Comment (mattcaswell) [reply crypto/evp/p_lib.c]: Renamed as suggested.

**Review by t8m — APPROVED** [review]
LGTM

**Comment by mattcaswell** [comment]
Pushed.

**Pull request closed without merging**